icelake/src/io_v2/functional_writer/equality_delta_writer.rs

```diff
@@ -0,0 +1,426 @@
+use arrow_ord::partition::partition;
+use arrow_row::{OwnedRow, RowConverter, Rows, SortField};
+use arrow_schema::SchemaRef;
+use arrow_select::filter;
+use itertools::Itertools;
+use std::collections::HashMap;
+
+use crate::{types::FieldProjector, Error, ErrorKind};
+
+pub const INSERT_OP: i32 = 1;
+pub const DELETE_OP: i32 = 2;
+
+///`EqualityDeltaWriter` is same as: https://github.com/apache/iceberg/blob/2e1ec5fde9e6fecfbc0883465a585a1dacb58c05/core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java#L108
+///
+///EqualityDeltaWriter will guarantee that there is only one row with the unique columns value written. When
+///insert a row with the same unique columns value, it will delete the previous row.
+///
+///NOTE:
+///This write is not as same with `upsert`. If the row with same unique columns is not written in
+///this writer, it will not delete it.
+#[derive(Clone)]
+pub struct EqualityDeltaWriterBuilder<
+    DB: IcebergWriterBuilder,
+    PDB: IcebergWriterBuilder,
+    EDB: IcebergWriterBuilder,
+> where
+    DB::R: SingleFileWriterStatus + IcebergWriter,
+    PDB::R: IcebergWriter<PositionDeleteInput>,
+{
+    data_file_writer_builder: DB,
+    sorted_cache_writer_builder: PDB,
+    equality_delete_writer_builder: EDB,
+    unique_column_ids: Vec<usize>,
+}
+
+impl<DB: IcebergWriterBuilder, PDB: IcebergWriterBuilder, EDB: IcebergWriterBuilder>
+    EqualityDeltaWriterBuilder<DB, PDB, EDB>
+where
+    DB::R: SingleFileWriterStatus + IcebergWriter,
+    PDB::R: IcebergWriter<PositionDeleteInput>,
+    EDB::R: IcebergWriter,
+{
+    pub fn new(
+        data_file_writer_builder: DB,
+        sorted_cache_writer_builder: PDB,
+        equality_delete_writer_builder: EDB,
+        unique_column_ids: Vec<usize>,
+    ) -> Self {
+        Self {
+            data_file_writer_builder,
+            sorted_cache_writer_builder,
+            equality_delete_writer_builder,
+            unique_column_ids,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<DB: IcebergWriterBuilder, PDB: IcebergWriterBuilder, EDB: IcebergWriterBuilder>
+    IcebergWriterBuilder for EqualityDeltaWriterBuilder<DB, PDB, EDB>
+where
+    DB::R: SingleFileWriterStatus + IcebergWriter,
+    PDB::R: IcebergWriter<PositionDeleteInput>,
+    EDB::R: IcebergWriter,
+{
+    type R = EqualityDeltaWriter<DB::R, PDB::R, EDB::R>;
+
+    async fn build(self, schema: &SchemaRef) -> Result<Self::R> {
+        let data_file_writer = self.data_file_writer_builder.build(schema).await?;
+        let position_delete_writer = self.sorted_cache_writer_builder.build(schema).await?;
+        let equality_delete_writer = self.equality_delete_writer_builder.build(schema).await?;
+        EqualityDeltaWriter::try_new(
+            data_file_writer,
+            position_delete_writer,
+            equality_delete_writer,
+            self.unique_column_ids,
+            schema,
+        )
+    }
+}
+
+pub struct EqualityDeltaWriter<
```

Comment:
```suggestion
pub struct DeltaWriter<
```

Reply:
Actually, we have the same name writer `DeltaWriter` in input_wrapper which expose the interface like `write(op,batch)`. 🤣  Here the EqualityDelta is follow [iceberg-java](https://github.com/apache/iceberg/blob/2e1ec5fde9e6fecfbc0883465a585a1dacb58c05/core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java#L108)